services/paymaster/src/constraint.rs

```diff
@@ -0,0 +1,148 @@
+use serde::{Serialize, Deserialize};
+use solana_pubkey::Pubkey;
+use utoipa::ToSchema;
+use serde_with::{serde_as, DisplayFromStr};
+
+#[derive(Serialize, Deserialize, ToSchema)]
+#[serde(tag = "version")]
+pub enum TransactionVariation {
+    #[serde(rename = "v1")]
+    #[schema(title = "v1")]
+    V1(TransactionVariationV1),
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct TransactionVariationV1 {
+	pub name: String,
+	pub instructions: Vec<InstructionConstraint>,
+	pub rate_limits: RateLimits,
+	pub max_gas_spend: u64,
+}
+
+
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct RateLimits {
+	user_per_min: u64,
+	session_per_min: u64,
+	ip_per_min: u64,
+}
```

Comment:
would be nice if we could track and enforce spending caps on these too, but idk if it's feasible